### ISSUE TITLE
Feature request: add weights to scipy.stats.linregress

### ISSUE THREAD

**Issue by tantrev** [scipy/scipy#7438]
Pretty simple: would just be nice if a weighting vector could be used with the scipy.stats.linregress function, similar to the "weights" option in Matlab's cftool.

**Comment by rgommers** [comment]
I'd say the trade-off is similar to https://github.com/scipy/scipy/issues/7399#issuecomment-302881575, although adding weights is a smaller extension. I'm +0.5 for this request.

**Comment by josef-pkt** [comment]
adding weights is a pretty simple and straight forward extension
+1

**Comment by tantrev** [comment]
@josef-pkt for those of us who aren't mathematically talented, do you by chance have any reference code for a simple implementation?

**Comment by josef-pkt** [comment]
no, not really, the main computation is the same but you multiply both arrays by the weights array before doing any of the computations. statsmodels has WLS to get the numbers for a unit test.

The last three returns should also just follow from the weighted variables, but those are not formulas that I have memorized because the standard OLS way is different.
(rsquared might have a different definition, but AFAICS, `r` here is the simple correlation.

**Comment by josef-pkt** [comment]
`optimize.curve_fit` does just the weighting of the y and the prediction function, IIRC. That might be another case to compare against.

It's easier to compute and unit test first and then figure out what's wrong if the results don't match. (Essentially I'm always getting confused about whether we need to square the weights for the mean or not.)

**Comment by tantrev** [comment]
Thank you!  Now I feel stupid for realizing it was that simple. :P

On the rsquared side of things, for what it's worth, Matlab seems to calculate a different "adjusted" rsquared when weighting is used.

When using Matlab's cftool with the following arrays:

```
a=[1 2 3 4 5];
b=[1 2 3 4 5]**7;
c=[9 8 7 6 5];
```

It kicks out the following calculation for a linear weighted fit:

> Linear model Poly1:
>      f(x) = p1*x + p2
> Coefficients (with 95% confidence bounds):
>        p1 =   1.523e+04  (-6372, 3.684e+04)
>        p2 =  -2.691e+04  (-9.275e+04, 3.893e+04)
> 
> Goodness of fit:
>   SSE: 9.285e+09
>   R-square: 0.6266
>   Adjusted R-square: 0.5022
>   RMSE: 5.563e+04

Personally, I just care about the p-value calculation being correct with the weights, but I suppose it's probably worth getting the rsquared correct as well.

**Comment by tantrev** [comment]
Also, it looks like Matlab also has an "adjusted" rsquared value when weights are not used at all (with the "Polynomial, degree 1" configuration in cftool):

Here's cftool's output when just fitting to just the a and b arrays:

> Linear model Poly1:
>      f(x) = p1*x + p2
> Coefficients (with 95% confidence bounds):
>        p1 =   1.725e+04  (-5447, 3.995e+04)
>        p2 =  -3.239e+04  (-1.077e+05, 4.289e+04)
> 
> Goodness of fit:
>   SSE: 1.526e+09
>   R-square: 0.661
>   Adjusted R-square: 0.548
>   RMSE: 2.255e+04

**Comment by josef-pkt** [comment]
adjusted r-square adjusts for the number of regressors. Given that there are only two parameters and no choice for more, adjusted rsquared is not available in scipy. Only rsquare is relevant for linregress.
